README.md
# Universal JSON Encoder/Decoder for Python Objects #

This small Python package enables encoding / decoding most Python objects into / from JSON. It offers an API similar to the `json` standard library package but provides automated functionalities to encode / decode most objects that cannot be converted into JSON by default (e.g. datetimes or custom classes). To do this, it adds two properties `__class__` and `__module__` to the generated JSON for all types normally not supported.

The rationale for this package is pretty simple: I like to be able to dump/load my objects in an easily readable format pretty much anywhere (e.g. a DB, another program, another Python program). This is particularly practical to use with applications using MongoDB.

This package works with Python version 2.7, 3.4, 3.5, 3.6 and 3.7 (the unit tests passed for all those versions). To install it, clone the git and simply use `python setup.py install` OR you can also install it via `pip install unijson` since it's also registered on PyPi [here](https://pypi.org/project/unijson/).

Here is what you can do with this little package.

## The types already supported by `json` are unchanged ##

All the types already supported by `json` are supported by `unijson` and behave in the exact same way since `UniversalJSONEncoder` uses the default `JSONEncoder` for all the types already supported.

```python
import json, unijson

d = {"peuh": 123, "pouet": [True, False, None, "ruguhregr"]}

# Both will return the same string:
json.dumps(d)
unijson.dumps(d)
```

## Automatically find a way to encode your custom classes ##

Using `unijson`, converting your custom classes instances into JSON objects has never been more simple.
For most simple classes, the `UniversalJSONEncoder` will find a way to transform them into JSON compliant objects without the need for extra methods.

```python
# Define a custom class
class Test(object):
    def __init__(self, a1, a2):
        self.a1, self.a2 = a1, a2

o1 = Test(12, 34)

# json.dumps(o1) would raise an exception.
# TypeError: Object of type 'Test' is not JSON serializable

# unijson won't blink:
unijson.dumps(o1) # {"a1": 12, "a2": 34, "__class__": "Test", "__module__": "examples"}
```

## Define encoders / decoders for your custom classes ##

With `unijson`, providing encoding / decoding functions for your custom classes is really simple. You just implement `__json_encode__()` / `__json_decode__()` in your custom classes. They will then be taken into account automatically by the `UniversalJSONEncoder` / `UniversalJSONDecoder`. This can be used to add some extra information into the resulting JSON object, or to provide a way to transform complex object into a JSON-serialisable dictionnary.

```python
# Define a custom class that defines methods for the universal serialiser:
class DefineMethods(object):
    def __init__(self, a1, a2):
        self.a1, self.a2 = a1, a2

    def __json_encode__(self):
        return {"a1":self.a1, "a2":self.a2, "extra":"for the fun"}

    @staticmethod
    def __json_decode__(d):
        return DefineMethods(d["a1"], d["a2"])

    def __eq__(self, other):
        return self.__dict__ == other.__dict__

o = DefineMethods("whatever", False)
s = unijson.dumps(o) # '{"a1": "whatever", "a2": false, "extra": "for the fun", "__class__": "DefineMethods", "__module__": "examples"}'
d = unijson.loads(s) # o == d should be True
```

Here is what you need to know on the two methods illustrated above:
* `__json_encode__()`: should take no argument and return a dictionary of unijson-serialisable objects (remember that many objects are unijson-serialisable by default!).
* `__json_decode__()`: should be static, take a single argument (a dict), and return an instance of the class it's defined in.

## Define and register encoders / decoders for any class ##

If you use external libraries, or non-JSON-serialisable types from the standard library (e.g. datetime), you might find yourself stuck on a simple way to convert those objects into JSON. Worry no more, my friend, for `unijson` has a solution for that. You can register encoding / decoding functions to the `UniversalJSONEncoder` / `UniversalJSONDecoder` to specify how to encode / decode anything. `unijson` already offers a few extra encoders / decoders for all the types found in `datetime` and `pytz` (for the management of timezones). If you have a look at `unijson`'s code, you will find those encoders. Here is an example that demonstrates how to register encoding functions:

```python
#Encoding function:
def json_encode_date(d):
    """Encoder for dates (from module datetime)."""
    return {"day"   : d.day,
            "month" : d.month,
            "year"  : d.year}
UniversalJSONEncoder.register(datetime.date, json_encode_date)
# Won't even require a decoding function since the created dictionnary can be used as arguments to the constructor (that's one of the automated methods used by UniversalJSONDecoder to build objects).
```

If you want to add your own encoders / decoders and register them, the functions you register must do the following:
* encoders should take single argument (the object to encode) and return a dictionary of UniJSON-serialisable objects.
* decoders should take a single argument (the dict extracted by the decoder) and return an instance of the decoded object.

# Additional information #

Author: Bastien Pietropaoli

Contact: bastien.pietropaoli@gmail.com

License: [Apache v2.0](http://www.apache.org/licenses/LICENSE-2.0)

If you have suggestions, remarks, or questions, do not hesitate to contact me.
Also, if you have made nice encoders/decoders that you think could be a good addition to this package, feel free to submit a pull request.


CHANGELOG.md
# unijson - Changelog #

The aim of this file is to track the various changes that have / will occur in the development of this package. This is supposed to be human readable. For that purpose, the [Keep a CHANGELOG](https://keepachangelog.com/en/1.0.0/) is used. Moreover, this package adheres to [Semantic Versioning](http://semver.org/).

# CHANGELOG

## [1.0.0] - 2018-08-13

First release of the package. Here is what if offers:
* A `json`-like API with the same `dumps()` / `loads()` functions with extended functionalities.
* A `UniversalJSONEncoder` and a `UniversalJSONDecoder` class for everyone to use and extend.
* An easy way to provide encoders / decoders by simply adding the `__json_encode__()` method and `__json_decode__()` static method to your classes.
* An easy way to register new encoders / decoders when working with standard library and external libraries types.


unittests/test_ujson.py
"""
Copyright (c) 2018 Bastien Pietropaoli

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from __future__ import absolute_import
import unittest
import datetime, pytz

# Relative import from parent directory as found here:
# https://gist.github.com/JungeAlexander/6ce0a5213f3af56d7369
import os,sys,inspect
current_dir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)

import ujson, json


#########################################################################################
#########################################################################################
#########################################################################################


# -------------
# Test classes:
# -------------

class NothingDefined(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


class DefineEncoder(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
    def __json_encode__(self):
        return {"a1":self.a1, "a2":self.a2, "source":"__json_encode__"}
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


class DefineDecoder(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
        self.source = ""
    @staticmethod
    def __json_decode__(d):
        o = DefineDecoder(**d)
        o.source = "__json_decode__"
        return o
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


class DefineBoth(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
    def __json_encode__(self):
        return {"a1":self.a1, "a2":self.a2}
    @staticmethod
    def __json_decode__(d):
        return DefineBoth(d["a1"], d["a2"])
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


#########################################################################################
#########################################################################################
#########################################################################################


class TestUJSON(unittest.TestCase):

    def test_encoder(self):
        # Test already supported types:
        d = {"peuh":12}
        expected = '{"peuh": 12}'
        self.assertEqual(ujson.dumps(d), expected)
        self.assertEqual(ujson.dumps(d), json.dumps(d))

        l = [False, True, None, ("Well", "I test thoroughly")]
        expected = '[false, true, null, ["Well", "I test thoroughly"]]'
        self.assertEqual(ujson.dumps(l), expected)
        self.assertEqual(ujson.dumps(l), json.dumps(l))

        # Test custom classes:
        o = NothingDefined(12, 55789)
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__": "test_ujson", "__class__": "NothingDefined", "a1": 12, "a2": 55789}
        self.assertDictEqual(d, expected)

        o = DefineEncoder(1643, NothingDefined("sgfd", [12, False, True, "peuh"]))
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__": "test_ujson", "__class__": "DefineEncoder", "a1": 1643,
                    "a2": {"__module__": "test_ujson", "__class__": "NothingDefined", "a1": "sgfd",
                            "a2": [12, False, True, "peuh"]}, "source":"__json_encode__"}
        self.assertDictEqual(d, expected)

        # Test datetime types: (also testing that registering methods actually works)
        o = pytz.timezone("UTC")
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"pytz", "__class__":"timezone", "zone":"UTC"}
        self.assertDictEqual(d, expected)

        o = datetime.date(2018, 8, 13)
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"date", "year":2018, "month":8, "day":13}
        self.assertDictEqual(d, expected)

        o = datetime.datetime(2018, 8, 13, 18, 53, 42, tzinfo=pytz.timezone("Europe/Dublin"))
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"datetime", "year":2018, "month":8, "day":13,
                    "hour":18, "minute":53, "second":42, "microsecond":0, "tzinfo":{"__module__":"pytz",
                    "__class__":"timezone", "zone":"Europe/Dublin"}}
        self.assertDictEqual(d, expected)

        o = datetime.time(18, 53, 42)
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"time", "hour":18, "minute":53, "second":42,
                    "microsecond":0, "tzinfo":None}
        self.assertDictEqual(d, expected)

        o = datetime.timedelta(seconds=3672)
        s = ujson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"timedelta", "seconds":3672}
        self.assertDictEqual(d, expected)


    def test_decoder(self):
        # Test already supported types:
        s = '{"peuh":12}'
        o = ujson.loads(s)
        expected = {"peuh":12}
        self.assertDictEqual(o, expected)
        self.assertDictEqual(o, json.loads(s))

        s = '[false, true, null, ["Well", "I test thoroughly"]]'
        o = ujson.loads(s)
        expected = [False, True, None, ["Well", "I test thoroughly"]]
        self.assertEqual(o, expected)
        self.assertEqual(o, json.loads(s))

        # Test custom classes:
        s = '{"__module__": "test_ujson", "__class__": "NothingDefined", "a1": 12, "a2": 55789}'
        o = ujson.loads(s)
        expected = NothingDefined(12, 55789)
        self.assertEqual(o, expected)

        s = '{"__module__":"test_ujson", "__class__":"DefineDecoder", "a1":12, "a2":55789}'
        o = ujson.loads(s)
        expected = DefineDecoder(12, 55789)
        expected.source = "__json_decode__"
        self.assertEqual(o, expected)

        s = '''{"a1": 1643, "a2": {"a1": "sgfd", "a2": [12, false, true, "peuh"], "__class__": "NothingDefined",
                "__module__": "test_ujson"}, "__class__": "DefineDecoder", "__module__": "test_ujson"}'''
        o = ujson.loads(s)
        expected = DefineDecoder(1643, NothingDefined("sgfd", [12, False, True, "peuh"]))
        expected.source = "__json_decode__"
        self.assertEqual(o, expected)

        # Test the datetime types:
        s = '{"__module__":"pytz", "__class__":"timezone", "zone":"UTC"}'
        o = ujson.loads(s)
        expected = pytz.timezone("UTC")
        self.assertEqual(o, expected)

        s = '{"__module__":"datetime", "__class__":"date", "year":2018, "month":8, "day":13}'
        o = ujson.loads(s)
        expected = datetime.date(2018, 8, 13)
        self.assertEqual(o, expected)

        s = '''{"__module__":"datetime", "__class__":"datetime", "year":2018, "month":8, "day":13,
                "hour":18, "minute":53, "second":42, "microsecond":0, "tzinfo":{"__module__":"pytz",
                "__class__":"timezone", "zone":"Europe/Dublin"}}'''
        o = ujson.loads(s)
        expected = datetime.datetime(2018, 8, 13, 18, 53, 42, tzinfo=pytz.timezone("Europe/Dublin"))

        s = '''{"__module__":"datetime", "__class__":"time", "hour":18, "minute":53, "second":42,
                "microsecond":0, "tzinfo":null}'''
        o = ujson.loads(s)
        expected = datetime.time(18, 53, 42)
        self.assertEqual(o, expected)

        s = '{"__module__":"datetime", "__class__":"timedelta", "seconds":3672}'
        o = ujson.loads(s)
        expected = datetime.timedelta(seconds=3672)
        self.assertEqual(o, expected)


    def test_serialiser(self):
        # Test equality of various objects before and after
        # a round of encoding and decoding.
        o = {"peuh": 12, "pouet":[False, True, None, ["Well", "I test thoroughly"]]}
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        o = NothingDefined(12, 55789)
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        o = DefineBoth("fduhghdgg", [None, "sdifj", 12, False, True])
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        for tz in pytz.all_timezones:
            o = pytz.timezone(tz)
            self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        o = datetime.date(2018, 8, 13)
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        o = datetime.datetime(2018, 8, 13, 18, 53, 42, tzinfo=pytz.timezone("Europe/Dublin"))
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        o = datetime.time(18, 53, 42)
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))

        o = datetime.timedelta(seconds=3672)
        self.assertEqual(o, ujson.loads(ujson.dumps(o)))


#########################################################################################
#########################################################################################
#########################################################################################


if __name__ == "__main__":
    unittest.main()


unittests/test_unijson.py
"""
Copyright (c) 2018 Bastien Pietropaoli

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

from __future__ import absolute_import
import unittest
import datetime, pytz

# Relative import from parent directory as found here:
# https://gist.github.com/JungeAlexander/6ce0a5213f3af56d7369
import os,sys,inspect
current_dir = os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
parent_dir = os.path.dirname(current_dir)
sys.path.insert(0, parent_dir)

import unijson, json


#########################################################################################
#########################################################################################
#########################################################################################


# -------------
# Test classes:
# -------------

class NothingDefined(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


class DefineEncoder(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
    def __json_encode__(self):
        return {"a1":self.a1, "a2":self.a2, "source":"__json_encode__"}
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


class DefineDecoder(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
        self.source = ""
    @staticmethod
    def __json_decode__(d):
        o = DefineDecoder(**d)
        o.source = "__json_decode__"
        return o
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


class DefineBoth(object):
    def __init__(self, a1, a2):
        self.a1 = a1
        self.a2 = a2
    def __json_encode__(self):
        return {"a1":self.a1, "a2":self.a2}
    @staticmethod
    def __json_decode__(d):
        return DefineBoth(d["a1"], d["a2"])
    def __eq__(self, other):
        return self.__dict__ == other.__dict__
    def __ne__(self, other):
        return not self.__eq__(other)


#########################################################################################
#########################################################################################
#########################################################################################


class Testunijson(unittest.TestCase):

    def test_encoder(self):
        # Test already supported types:
        d = {"peuh":12}
        expected = '{"peuh": 12}'
        self.assertEqual(unijson.dumps(d), expected)
        self.assertEqual(unijson.dumps(d), json.dumps(d))

        l = [False, True, None, ("Well", "I test thoroughly")]
        expected = '[false, true, null, ["Well", "I test thoroughly"]]'
        self.assertEqual(unijson.dumps(l), expected)
        self.assertEqual(unijson.dumps(l), json.dumps(l))

        # Test custom classes:
        o = NothingDefined(12, 55789)
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__": "test_unijson", "__class__": "NothingDefined", "a1": 12, "a2": 55789}
        self.assertDictEqual(d, expected)

        o = DefineEncoder(1643, NothingDefined("sgfd", [12, False, True, "peuh"]))
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__": "test_unijson", "__class__": "DefineEncoder", "a1": 1643,
                    "a2": {"__module__": "test_unijson", "__class__": "NothingDefined", "a1": "sgfd",
                            "a2": [12, False, True, "peuh"]}, "source":"__json_encode__"}
        self.assertDictEqual(d, expected)

        # Test datetime types: (also testing that registering methods actually works)
        o = pytz.timezone("UTC")
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"pytz", "__class__":"timezone", "zone":"UTC"}
        self.assertDictEqual(d, expected)

        o = datetime.date(2018, 8, 13)
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"date", "date":"2018-08-13"}
        self.assertDictEqual(d, expected)

        o = datetime.datetime(2018, 8, 13, 18, 53, 42, tzinfo=pytz.timezone("Europe/Dublin"))
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"datetime", "datetime":"2018-08-13 18:53:42.000000",
                    "tzinfo":{"__module__":"pytz", "__class__":"timezone", "zone":"Europe/Dublin"}}
        self.assertDictEqual(d, expected)

        o = datetime.time(18, 53, 42)
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"time", "time": "18:53:42.000000", "tzinfo":None}
        self.assertDictEqual(d, expected)

        o = datetime.timedelta(seconds=3672)
        s = unijson.dumps(o)
        d = json.loads(s)
        expected = {"__module__":"datetime", "__class__":"timedelta", "seconds":3672}
        self.assertDictEqual(d, expected)


    def test_decoder(self):
        # Test already supported types:
        s = '{"peuh":12}'
        o = unijson.loads(s)
        expected = {"peuh":12}
        self.assertDictEqual(o, expected)
        self.assertDictEqual(o, json.loads(s))

        s = '[false, true, null, ["Well", "I test thoroughly"]]'
        o = unijson.loads(s)
        expected = [False, True, None, ["Well", "I test thoroughly"]]
        self.assertEqual(o, expected)
        self.assertEqual(o, json.loads(s))

        # Test custom classes:
        s = '{"__module__": "test_unijson", "__class__": "NothingDefined", "a1": 12, "a2": 55789}'
        o = unijson.loads(s)
        expected = NothingDefined(12, 55789)
        self.assertEqual(o, expected)

        s = '{"__module__":"test_unijson", "__class__":"DefineDecoder", "a1":12, "a2":55789}'
        o = unijson.loads(s)
        expected = DefineDecoder(12, 55789)
        expected.source = "__json_decode__"
        self.assertEqual(o, expected)

        s = '''{"a1": 1643, "a2": {"a1": "sgfd", "a2": [12, false, true, "peuh"], "__class__": "NothingDefined",
                "__module__": "test_unijson"}, "__class__": "DefineDecoder", "__module__": "test_unijson"}'''
        o = unijson.loads(s)
        expected = DefineDecoder(1643, NothingDefined("sgfd", [12, False, True, "peuh"]))
        expected.source = "__json_decode__"
        self.assertEqual(o, expected)

        # Test the datetime types:
        s = '{"__module__":"pytz", "__class__":"timezone", "zone":"UTC"}'
        o = unijson.loads(s)
        expected = pytz.timezone("UTC")
        self.assertEqual(o, expected)

        s = '{"__module__":"datetime", "__class__":"date", "date":"2018-08-13"}'
        o = unijson.loads(s)
        expected = datetime.date(2018, 8, 13)
        self.assertEqual(o, expected)

        s = '''{"__module__":"datetime", "__class__":"datetime", "datetime":"2018-08-13 18:53:42.000000",
                "tzinfo":{"__module__":"pytz", "__class__":"timezone", "zone":"Europe/Dublin"}}'''
        o = unijson.loads(s)
        expected = datetime.datetime(2018, 8, 13, 18, 53, 42, tzinfo=pytz.timezone("Europe/Dublin"))

        s = '''{"__module__":"datetime", "__class__":"time", "time":"18:53:42.000000", "tzinfo":null}'''
        o = unijson.loads(s)
        expected = datetime.time(18, 53, 42)
        self.assertEqual(o, expected)

        s = '{"__module__":"datetime", "__class__":"timedelta", "seconds":3672}'
        o = unijson.loads(s)
        expected = datetime.timedelta(seconds=3672)
        self.assertEqual(o, expected)


    def test_serialiser(self):
        # Test equality of various objects before and after
        # a round of encoding and decoding.
        o = {"peuh": 12, "pouet":[False, True, None, ["Well", "I test thoroughly"]]}
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        o = NothingDefined(12, 55789)
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        o = DefineBoth("fduhghdgg", [None, "sdifj", 12, False, True])
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        for tz in pytz.all_timezones:
            o = pytz.timezone(tz)
            self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        o = datetime.date(2018, 8, 13)
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        o = datetime.datetime(2018, 8, 13, 18, 53, 42, tzinfo=pytz.timezone("Europe/Dublin"))
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        o = datetime.time(18, 53, 42)
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))

        o = datetime.timedelta(seconds=3672)
        self.assertEqual(o, unijson.loads(unijson.dumps(o)))


#########################################################################################
#########################################################################################
#########################################################################################


if __name__ == "__main__":
    unittest.main()


examples.py
'''
Copyright (c) 2018 Bastien Pietropaoli

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
'''

import json, unijson


###################################################################################################
# unijson for already supported types:
d = {"peuh": 123, "pouet": [True, False, None, "ruguhregr"]}

# Both will return the same string:
json.dumps(d)
unijson.dumps(d)


###################################################################################################
# unijson for custom classes:

# Define a custom class
class Test(object):
    def __init__(self, a1, a2):
        self.a1, self.a2 = a1, a2

o1 = Test(12, 34)

# json.dumps(o1) would raise an exception.
# TypeError: Object of type 'Test' is not JSON serializable

# unijson won't blink:
unijson.dumps(o1) # {"a1": 12, "a2": 34, "__class__": "Test", "__module__": "examples"}


###################################################################################################
# Define __json_encode__() and __json_decode__():

# Define a custom class that defines methods for the universal serialiser:
class DefineMethods(object):
    def __init__(self, a1, a2):
        self.a1, self.a2 = a1, a2

    def __json_encode__(self):
        return {"a1":self.a1, "a2":self.a2, "extra":"for the fun"}

    @staticmethod
    def __json_decode__(d):
        return DefineMethods(d["a1"], d["a2"])

    def __eq__(self, other):
        return self.__dict__ == other.__dict__

o = DefineMethods("whatever", False)
s = unijson.dumps(o) # '{"a1": "whatever", "a2": false, "extra": "for the fun", "__class__": "DefineMethods", "__module__": "examples"}'
d = unijson.loads(s) # o == d should be True


###################################################################################################
# Register a new encoder:

def json_encode_date(d):
    """Encoder for dates (from module datetime)."""
    return {"day"   : d.day,
            "month" : d.month,
            "year"  : d.year}
UniversalJSONEncoder.register(datetime.date, json_encode_date)
